File: src/gateway/middleware/error_handler.py
# This file is deprecated. Use error_handler_new.py instead.

from .error_handler_new import setup_error_handlers
from typing import Dict, Any


def _infer_error_type(status_code: int, detail: str) -> str:
    """Infer error type from status code and detail message."""
    detail_lower = detail.lower()
    
    if "unauthorized" in detail_lower or "auth" in detail_lower:
        return "unauthorized"
    if "expired" in detail_lower and "trial" in detail_lower:
        return "trial-expired"
    if "limit" in detail_lower and "trial" in detail_lower:
        return "trial-limit-exceeded"
    if "approval" in detail_lower or "governor" in detail_lower:
        return "approval-required"
    if "budget" in detail_lower:
        return "budget-exceeded"
    if "permission" in detail_lower or "forbidden" in detail_lower:
        return "permission-denied"
    if "not found" in detail_lower:
        return "not-found"
    if "timeout" in detail_lower:
        return "gateway-timeout"
    if "unavailable" in detail_lower:
        return "service-unavailable"
    
    status_to_type = {
        401: "unauthorized", 403: "permission-denied", 404: "not-found",
        402: "budget-exceeded", 429: "trial-limit-exceeded",
        503: "service-unavailable", 504: "gateway-timeout", 500: "internal-server-error"
    }
    
    return status_to_type.get(status_code, "internal-server-error")


def _get_title_for_status(status_code: int) -> str:
    """Get human-readable title for HTTP status code."""
    titles = {
        400: "Bad Request", 401: "Unauthorized", 402: "Payment Required",
        403: "Forbidden", 404: "Not Found", 409: "Conflict",
        429: "Too Many Requests", 500: "Internal Server Error",
        502: "Bad Gateway", 503: "Service Unavailable", 504: "Gateway Timeout"
    }
    return titles.get(status_code, "Error")


def create_problem_details(error_type: str, status: int, detail: str, instance: str, **extra_fields) -> Dict[str, Any]:
    """Helper function to create RFC 7807 Problem Details dict."""
    problem = {
        "type": f"https://waooaw.com/errors/{error_type}",
        "title": _get_title_for_error_type(error_type),
        "status": status,
        "detail": detail,
        "instance": instance
    }
    problem.update(extra_fields)
    return problem


def _get_title_for_error_type(error_type: str) -> str:
    """Get title for error type."""
    titles = {
        "unauthorized": "Unauthorized", "invalid-token-format": "Invalid Token Format",
        "token-expired": "Token Expired", "invalid-token": "Invalid Token",
        "permission-denied": "Permission Denied", "trial-expired": "Trial Expired",
        "trial-limit-exceeded": "Trial Limit Exceeded", "approval-required": "Approval Required",
        "budget-exceeded": "Budget Exceeded", "not-found": "Not Found",
        "validation-error": "Validation Error", "conflict": "Conflict",
        "service-unavailable": "Service Unavailable", "gateway-timeout": "Gateway Timeout",
        "internal-server-error": "Internal Server Error"
    }
    return titles.get(error_type, "Error")


# Backwards compatibility
# Removed deprecated ErrorHandlingMiddleware class

File: src/mobile/src/lib/errorHandler.ts
/**
 * API Error Handler
 * Centralized error handling for API requests
 */

import { AxiosError } from 'axios';
import {
  APIError,
  NetworkError,
  TimeoutError,
  ProblemDetails,
} from '../types/api.types';

/**
 * Error severity levels
 */
export type ErrorSeverity = 'info' | 'warning' | 'error' | 'critical';

/**
 * Handled error result
 */
export interface HandledError {
  message: string;
  severity: ErrorSeverity;
  originalError: Error;
  shouldRetry: boolean;
  userMessage: string;
}

/**
 * Check if error is an Axios error
 */
function isAxiosError(error: unknown): error is AxiosError {
  return (error as AxiosError).isAxiosError === true;
}

/**
 * Check if error response is Problem Details format
 */
function isProblemDetails(data: unknown): data is ProblemDetails {
  return (
    typeof data === 'object' &&
    data !== null &&
    'type' in data &&
    'title' in data &&
    'status' in data &&
    'detail' in data
  );
}

/**
 * Handle API errors with RFC 7807 Problem Details support
 */
export function handleAPIError(error: unknown): HandledError {
  // Network/timeout errors
  if (error instanceof Error && error.message.includes('Network request failed')) {
    return {
      message: 'Network connection failed',
      severity: 'error',
      originalError: new NetworkError(),
      shouldRetry: true,
      userMessage: 'Unable to connect to server. Please check your internet connection.',
    };
  }

  if (error instanceof Error && error.message.includes('timeout')) {
    return {
      message: 'Request timeout',
      severity: 'warning',
      originalError: new TimeoutError(),
      shouldRetry: true,
      userMessage: 'Request took too long. Please try again.',
    };
  }

  // Axios errors
  if (isAxiosError(error)) {
    const status = error.response?.status;
    const data = error.response?.data;

    // RFC 7807 Problem Details
    if (isProblemDetails(data)) {
      const apiError = new APIError(data);
      return handleStatusCode(status || 500, apiError, data.detail);
    }

    // Generic status code handling
    return handleStatusCode(
      status || 500,
      error,
      error.message || 'An error occurred'
    );
  }

  // Unknown errors
  if (error instanceof Error) {
    return {
      message: error.message,
      severity: 'error',
      originalError: error,
      shouldRetry: false,
      userMessage: 'An unexpected error occurred. Please try again.',
    };
  }

  // Fallback
  return {
    message: 'Unknown error',
    severity: 'error',
    originalError: new Error(String(error)),
    shouldRetry: false,
    userMessage: 'Something went wrong. Please try again.',
  };
}

/**
 * Handle specific HTTP status codes
 */
function handleStatusCode(
  status: number,
  error: Error,
  detail: string
): HandledError {
  switch (status) {
    case 400:
      return {
        message: `Bad Request: ${detail}`,
        severity: 'warning',
        originalError: error,
        shouldRetry: false,
        userMessage: detail || 'Invalid request. Please check your input.',
      };

    case 401:
      return {
        message: 'Unauthorized - token expired or invalid',
        severity: 'warning',
        originalError: error,
        shouldRetry: false,
        userMessage: 'Your session has expired. Please sign in again.',
      };

    case 403:
      return {
        message: 'Forbidden - insufficient permissions',
        severity: 'warning',
        originalError: error,
        shouldRetry: false,
        userMessage: 'You do not have permission to access this resource.',
      };

    case 404:
      return {
        message: `Not Found: ${detail}`,
        severity: 'info',
        originalError: error,
        shouldRetry: false,
        userMessage: 'The requested resource was not found.',
      };

    case 409:
      return {
        message: `Conflict: ${detail}`,
        severity: 'warning',
        originalError: error,
        shouldRetry: false,
        userMessage: detail || 'A conflict occurred. Please try again.',
      };

    case 422:
      return {
        message: `Validation Error: ${detail}`,
        severity: 'warning',
        originalError: error,
        shouldRetry: false,
        userMessage: detail || 'Validation failed. Please check your input.',
      };

    case 429:
      return {
        message: 'Too Many Requests - rate limit exceeded',
        severity: 'warning',
        originalError: error,
        shouldRetry: true,
        userMessage: 'Too many requests. Please wait a moment and try again.',
      };

    case 500:
      return {
        message: 'Internal Server Error',
        severity: 'error',
        originalError: error,
        shouldRetry: true,
        userMessage: 'Server error occurred. Please try again later.',
      };

    case 502:
    case 503:
    case 504:
      return {
        message: 'Service Unavailable',
        severity: 'error',
        originalError: error,
        shouldRetry: true,
        userMessage: 'Service is temporarily unavailable. Please try again later.',
      };

    default:
      return {
        message: `HTTP ${status}: ${detail}`,
        severity: status >= 500 ? 'error' : 'warning',
        originalError: error,
        shouldRetry: status >= 500,
        userMessage: detail || 'An error occurred. Please try again.',
      };
  }
}

/**
 * Log error to console (development) or analytics (production)
 */
export function logError(error: HandledError, context?: string): void {
  const prefix = context ? `[${context}]` : '[API Error]';
  
  if (__DEV__) {
    console.error(prefix, {
      message: error.message,
      severity: error.severity,
      shouldRetry: error.shouldRetry,
      userMessage: error.userMessage,
      originalError: error.originalError,
    });
  } else {
    // In production, send to analytics/monitoring service
    // Example: Sentry, Firebase Crashlytics, etc.
    console.warn(prefix, error.message);
  }
}

/**
 * Get user-friendly error message
 */
export function getUserMessage(error: unknown): string {
  const handled = handleAPIError(error);
  return handled.userMessage;
}

/**
 * Check if error should trigger retry
 */
export function shouldRetry(error: unknown): boolean {
  const handled = handleAPIError(error);
  return handled.shouldRetry;
}
